src/main/java/uk/ac/ebi/ampt2d/accession/AccessioningService.java

```diff
@@ -31,21 +32,20 @@ public AccessioningService(AccessionRepository<T> accessionRepository,
     }
 
     public Map<T, String> createAccessions(List<T> objects) {
-        Map<T, String> accessions = new HashMap<>();
-
-        for (T object : objects) {
-            // look for an existing accession for this object
-            String accession = accessionRepository.get(object);
-
-            if (accession == null) {
-                // if there is no previous accession for the object, create a new one and store it in the repository
-                accession = accessionGenerator.get(object);
-                accessionRepository.add(object, accession);
-            }
-
-            accessions.put(object, accession);
+        // look for accessions for those objects in the repository
+        Map<T, String> storedAccessions = accessionRepository.get(objects);
+
+        // get all objects that are not in the repository
+        Set<T> objectsNotInRepository = objects.stream().filter(object -> !storedAccessions.containsKey(object))
+                                           .collect(Collectors.toSet());
```

Comment:
If you are going to do it this way, then you can use the foreach in the stream.

Anyway probably the put operation should also be a Map<T,String> (or a list of Map.Entry) so that any underlying method can do batched writes. Otherwise we will hit performance when using databases.